src/ThgHostingClient.php

```diff
@@ -0,0 +1,539 @@
+        }
+
+        $requestParams = $this->allowedMethods[$method];
+        $headers = [
+            "X-Api-Token: " . $this->xApiToken,
+            "Content-Type: " . $contentType,
+            "Accept: application/json"
+        ];
+        $curl = curl_init();
+
+        curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($curl, CURLOPT_TIMEOUT, $this->timeout);
+        curl_setopt($curl, CURLOPT_FOLLOWLOCATION, TRUE);
+        curl_setopt($curl, ...$requestParams);
+
+        $url = $this->host . trim($endpoint, '/') . '/';
+
+        if (\sizeof($files) > 0) {
+            $arguments['attachments'] = [];
+            foreach ($files as $file) {
+                // Path to the file
+                if (\is_string($file) && \is_file($file)) {
+                    $stream = file_get_contents($file);
+                    $arguments['attachments'][] = [
+                        "file" => base64_encode($stream),
+                        "mime"    => mime_content_type($file),
+                        "name"   => basename($file)
+                    ];
+                    continue;
+                }
+                // File encoded in base64 with all required informations
+                if (\is_array($file)) {
+                    if (!isset($file["file"])) {
+                        throw new ThgHostingException("File encoded into base64 was not found", 404);
+                    }
+
+                    if (!isset($file["name"])) {
+                        throw new ThgHostingException("Name of the file was not found", 404);
+                    }
+
+                    if (!isset($file["mime"])) {
+                        throw new ThgHostingException("Mime type of file was not found", 404);
+                    }
+
+                    $arguments['attachments'][] = [
+                        "file" => $file["file"],
+                        "mime"    => $file["mime"],
+                        "name"   => $file["name"]
+                    ];
+                    continue;
+                }
+
+                throw new ThgHostingException("Passed file wasn't a path or a stream, couldn't be send - cancelling request.", 400);
+            }
+        }
+
+        if ($method === self::GET && \sizeof($arguments) > 0) {
+            $url .= "?" . http_build_query($arguments);
+        } else {
+            $arguments = json_encode($arguments);
+            curl_setopt($curl, CURLOPT_POSTFIELDS, $arguments);
+            $headers[] = 'Content-Length: ' . strlen($arguments);
+        }
+        curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
+
+        curl_setopt($curl, CURLOPT_URL, $url);
+        $result = curl_exec($curl);
+        curl_close($curl);
+
+        return [
+            "data" => gettype($result) == 'string' ? (json_decode($result, true) ?? $result) : $result,
+            "info" => curl_getinfo($curl)
```

Comment:
I can't find any difference between getting info after I close the request or before:
Before:
```
Info:

    (string) url => `http://gateway.localhost/rest-api/servers/24/`
    (string) content_type => `application/json`
    (integer) http_code => `400`
    (integer) header_size => `537`
    (integer) request_size => `221`
    (integer) filetime => `-1`
    (integer) ssl_verify_result => `0`
    (integer) redirect_count => `0`
    (double) total_time => `0.980218`
    (double) namelookup_time => `0.000544`
    (double) connect_time => `0.000637`
    (double) pretransfer_time => `0.000675`
    (double) size_upload => `2`
    (double) size_download => `173`
    (double) speed_download => `176`
    (double) speed_upload => `2`
    (double) download_content_length => `-1`
    (double) upload_content_length => `2`
    (double) starttransfer_time => `0.980194`
    (double) redirect_time => `0`
    (string) redirect_url => ``
    (string) primary_ip => `127.0.0.1`
    (array) [0] certinfo:
        [EMPTY]
    (integer) primary_port => `80`
    (string) local_ip => `127.0.0.1`
    (integer) local_port => `56186`
    (integer) http_version => `2`
    (integer) protocol => `1`
    (integer) ssl_verifyresult => `0`
    (string) scheme => `HTTP`
    (integer) appconnect_time_us => `0`
    (integer) connect_time_us => `637`
    (integer) namelookup_time_us => `544`
    (integer) pretransfer_time_us => `675`
    (integer) redirect_time_us => `0`
    (integer) starttransfer_time_us => `980194`
    (integer) total_time_us => `980218`
```
After:
```
Info:

    (string) url => `http://gateway.localhost/rest-api/servers/24/`
    (string) content_type => `application/json`
    (integer) http_code => `400`
    (integer) header_size => `537`
    (integer) request_size => `221`
    (integer) filetime => `-1`
    (integer) ssl_verify_result => `0`
    (integer) redirect_count => `0`
    (double) total_time => `0.874356`
    (double) namelookup_time => `0.000731`
    (double) connect_time => `0.000816`
    (double) pretransfer_time => `0.000847`
    (double) size_upload => `2`
    (double) size_download => `173`
    (double) speed_download => `197`
    (double) speed_upload => `2`
    (double) download_content_length => `-1`
    (double) upload_content_length => `2`
    (double) starttransfer_time => `0.87433`
    (double) redirect_time => `0`
    (string) redirect_url => ``
    (string) primary_ip => `127.0.0.1`
    (array) [0] certinfo:
        [EMPTY]
    (integer) primary_port => `80`
    (string) local_ip => `127.0.0.1`
    (integer) local_port => `56190`
    (integer) http_version => `2`
    (integer) protocol => `1`
    (integer) ssl_verifyresult => `0`
    (string) scheme => `HTTP`
    (integer) appconnect_time_us => `0`
    (integer) connect_time_us => `816`
    (integer) namelookup_time_us => `731`
    (integer) pretransfer_time_us => `847`
    (integer) redirect_time_us => `0`
    (integer) starttransfer_time_us => `874330`
    (integer) total_time_us => `874356`
```
But sure, I will implement it.

Reply:
I was getting an error that bool was passed instead of an array. curl_info was returning false

Reply:
Ah, this is not this fault. The response couldn't get properly decoded. I tried to create universal check to see if response it corrects and can be turned to json:
```php
$message = ($res['data']['message'] ?? null) ?? ($res['data']['error']['description'] ?? $res['data']);
```
But it still sometimes fails if something unexpected got send. Like the message is just `false`. It's not really expected when it happens, so I just left it because this is only for testing, and you can just go here and `var_dump` the response.